File: RestfulObjects Server/RestfulObjects.Test.Data/WithReference.cs
// Copyright © Naked Objects Group Ltd ( http://www.nakedobjects.net). 
// All Rights Reserved. This code released under the terms of the 
// Microsoft Public License (MS-PL) ( http://opensource.org/licenses/ms-pl.html) 

using System.ComponentModel.DataAnnotations;
using System.Linq;
using NakedObjects;
using Do = NakedObjects.EagerlyAttribute.Do;

namespace RestfulObjects.Test.Data {
    public class WithReference {
        public IDomainObjectContainer Container { set; protected get; }


        [Key, Title]
        public virtual int Id { get; set; }


        public virtual MostSimple AReference { get; set; }

        [Optionally]
        public virtual MostSimple ANullReference { get; set; }

        [Disabled]
        public virtual MostSimple ADisabledReference { get; set; }

        [Hidden]
        [Optionally]
        public virtual MostSimple AHiddenReference { get; set; }


        public virtual MostSimple AChoicesReference { get; set; }

        public virtual MostSimple[] ChoicesAChoicesReference() {
            return Container.Instances<MostSimple>().Where(ms => ms.Id == 1 || ms.Id == 2).ToArray();
        }

        public virtual MostSimple AnAutoCompleteReference { get; set; }

        public virtual IQueryable<MostSimple> AutoCompleteAnAutoCompleteReference([MinLength(2)]string name) {
            return Container.Instances<MostSimple>().Where(ms => name.Contains(ms.Id.ToString()));
        }

        public virtual string Validate(MostSimple aReference, MostSimple aChoicesReference) {
            if (aReference != null && aReference.Id == 1 && aChoicesReference.Id == 2) {
                return "Cross validation failed";
            }
            return "";
        }

        public virtual MostSimple AConditionalChoicesReference { get; set; }

        public virtual MostSimple[] ChoicesAConditionalChoicesReference(MostSimple aReference) {
            return Container.Instances<MostSimple>().Where(ms => ms.Id != aReference.Id).ToArray();
        }

        [Eagerly(Do.Rendering)]
        public virtual MostSimple AnEagerReference { get; set; }
    }
}

File: Rest/NakedObjects.Rest.Test.Data.Sdm2/WithReference.vb
' Copyright © Naked Objects Group Ltd ( http://www.nakedobjects.net). 
' All Rights Reserved. This code released under the terms of the 
' Microsoft Public License (MS-PL) ( http://opensource.org/licenses/ms-pl.html) 
Imports sdm.systems.application.control

Namespace RestfulObjects.Test.Data
    Public Class WithReference
        Inherits AbstractBusinessObject

#Region "Id and Title"

        Public Overrides Function title() As Title
            Dim t As New Title
            t.append(Id)
            Return t
        End Function

        Private _id As New WholeNumber

        Public ReadOnly Property Id() As WholeNumber
            Get
                Return _id
            End Get
        End Property

#End Region

#Region "AReference"

        Private myAReference As MostSimple

        Public Overridable Property AReference() As MostSimple
            Get
                resolve(MethodBase.GetCurrentMethod)
                Return myAReference
            End Get
            Set(ByVal value As MostSimple)
                resolve(MethodBase.GetCurrentMethod)
                myAReference = value
                objectChanged()
            End Set
        End Property

#End Region

#Region "ANullReference"

        Private myANullReference As MostSimple

        Public Overridable Property ANullReference() As MostSimple
            Get
                resolve(MethodBase.GetCurrentMethod)
                Return myANullReference
            End Get
            Set(ByVal value As MostSimple)
                resolve(MethodBase.GetCurrentMethod)
                myANullReference = value
                objectChanged()
            End Set
        End Property

#End Region


#Region "ADisabledReference"

        Private myADisabledReference As MostSimple

        Public Overridable Property ADisabledReference() As MostSimple
            Get
                resolve(MethodBase.GetCurrentMethod)
                Return myADisabledReference
            End Get
            Set(ByVal value As MostSimple)
                resolve(MethodBase.GetCurrentMethod)
                myADisabledReference = value
                objectChanged()
            End Set
        End Property

        Public Overridable Sub aboutADisabledReference(ByVal a As FieldAbout, ByVal value As MostSimple)
            a.unmodifiable("Field not editable")
        End Sub

#End Region

#Region "AHiddenReference"

        Private myAHiddenReference As MostSimple

        Public Overridable Property AHiddenReference() As MostSimple
            Get
                resolve(MethodBase.GetCurrentMethod)
                Return myAHiddenReference
            End Get
            Set(ByVal value As MostSimple)
                resolve(MethodBase.GetCurrentMethod)
                myAHiddenReference = value
                objectChanged()
            End Set
        End Property

        Public Overridable Sub aboutAHiddenReference(ByVal a As FieldAbout, ByVal value As MostSimple)
            a.invisible()
        End Sub

#End Region

#Region "AChoicesReference"

        'RP:  Choices exist for parameters, not for properties.
        'Public Overridable Property AChoicesValue() As Integer

        Private myAChoicesReference As MostSimple

        Public Overridable Property AChoicesReference() As MostSimple
            Get
                resolve(MethodBase.GetCurrentMethod)
                Return myAChoicesReference
            End Get
            Set(ByVal value As MostSimple)
                resolve(MethodBase.GetCurrentMethod)
                myAChoicesReference = value
                objectChanged()
            End Set
        End Property

        'Public Overridable Function ChoicesAChoicesReference() As MostSimple()
        '    'Return Container.Instances(Of MostSimple)().ToArray()
        'End Function

#End Region

#Region "Set up (Fixtures)"

        Public Shared Function SetUp(id As Integer, ref As MostSimple, dis As MostSimple) As WithReference
            Dim wr As WithReference =
                    CType(
                        createTransientInstanceFromShared(
                            GetType(WithReference)),
                        WithReference)

            wr.Id.setValue(id)
            wr.AReference = ref
            wr.ADisabledReference = dis
            wr.AChoicesReference = dis
            wr.actionSave()
            Return wr
        End Function

#End Region
    End Class
End Namespace
